### DISCUSSION TITLE
Functional testing: Start multiple test workers with different queues

### DISCUSSION THREAD

**Discussion by Brontes** [celery/celery#7711] [Q&A]
I am trying to do functional tests with multiple celery workers that operate in different queues.

I am using celery.contrib.testing.worker.start_worker function with param "queues" to start workers. I want to one task execute only in worker with queue "special" and other tasks in other workers.

But it seems to me that there is a bug. And if there are workers with different queues all tasks are executed by all workers.
If I create only workers for one queue it seems to work as it should - task from queue that doesn't have worker is not executed.

I have created minimum test case where this issue can be seen: https://github.com/Brontes/CeleryTestCase
I have three tasks: quick_task, slow_task and special task. 
Quick task gets executed very often (like simulating celery beat). And from time to time this quick task generates slow_task (which takes 3 seconds to execute) and/or special_task that should be executed by special worker.
But all tasks are executed by all workers.
And if I create only workers for one queue only tasks for this workers are executed. For instance: If all workers are for default queue only quick_task and slow_task will get executed... but special_task wont. As it should be. 
There is only problem if there are some workers for default queue and some for special queue. In this case all tasks are executed by all workers.

Am I doing something wrong or is there really a bug?

I also noticed some othere issues (occasionally) when deailng with this tests (but those can be discussed in other discussions):
- Workers get stale and stops receiving tasks
- "dictionary changed size during iteration" exeption during test execution that results in staling workers
- strange delegation of tasks for workers: One worker gets a lot of tasks in a row even if other worker are not doing anything and tasks are waiting for execution.

Any suggestions are really welcome.

**Comment by auvipy** [comment]
there are of course bugs there, that is why a member of our team is working on an enhanced testing framework https://github.com/celery/pytest-celery, you can check the open PR's

**Reply by Brontes** [reply]
Thank you for pointing me to this.

So do I understand correctly, that there is currently no stable/reliable/recommended way of doing (fairly complex) functional testing for projects that implements celery?
Or there is, and I just tried to do it the wrong way?

**Comment by Brontes** [comment]
Sorry to be a bit pushy here... 

But I would really like to know what is the best way for functional testing of projects that implements celery.
Any hints appreciated.
